File: app/ui/PlayEnvironment.java
package ui;

import com.google.inject.AbstractModule;
import com.google.inject.Guice;
import models.*;
import modules.IArchiveRepository;
import modules.IEnvironment;
import modules.IOrderModule;
import modules.ISearchModule;


/**
 * Created by Beemen on 30/07/2015.
 */
public class PlayEnvironment
        implements IEnvironment {

    @Override
    public ISearchModule SearchModule() {
        return GetObject(ISearchModule.class);
    }

    private IOrderModule _OrderModule;
    private IArchiveRepository _ArchiveRepository;

    @Override
    public IOrderModule OrderModule() {
        if(_OrderModule == null)
            _OrderModule = GetObject(IOrderModule.class);
        return _OrderModule;
    }

    @Override
    public IArchiveRepository ArchiveRepository() {
        if(_ArchiveRepository == null)
            _ArchiveRepository = GetObject(IArchiveRepository.class);
        return _ArchiveRepository;
    }

    @Override
    public ISession Session() {
        return GetObject(ISession.class);
    }

    private com.google.inject.Injector injector;
    public PlayEnvironment(){

        injector = Guice.createInjector(new AbstractModule() {
            @Override
            protected void configure() {
                bind(ISession.class).to(PlaySession.class);
                bind(ISearchModule.class).to(modules.FlatLilySearchModule.class);
                bind(IOrderModule.class).to(models.dummy.OrderModule.class);
                bind(IArchiveRepository.class).to(models.dummy.ArchiveRepository.class);
            }
        });
    }

    protected <T> T GetObject(Class<T> type)
    {
        return injector.<T>getInstance(type);
    }
}

File: UI/PlayEnvironment.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using end_user_gui.Models;
using end_user_gui.Modules;

namespace end_user_gui.UI
{
    public class PlayEnvironment : IEnvironment
    {

        public ISearchModule SearchModule()
        {
            return GetObject<ISearchModule>();
        }

        private IOrderModule _OrderModule;
        private IArchiveRepository _ArchiveRepository;

        public IOrderModule OrderModule()
        {
            if (_OrderModule == null)
                _OrderModule = GetObject<IOrderModule>();
            return _OrderModule;
        }

        public IArchiveRepository ArchiveRepository()
        {
            if (_ArchiveRepository == null)
                _ArchiveRepository = GetObject<IArchiveRepository>();
            return _ArchiveRepository;
        }

        public ISession Session()
        {
            return GetObject<ISession>();
        }

        public PlayEnvironment()
        {
            // TODO: put real injection here
            _Types[typeof(ISession)] = typeof(PlaySession);
            _Types[typeof(ISearchModule)] = typeof(Modules.FlatLilySearchModule);
            _Types[typeof(IOrderModule)] = typeof(Models.dummy.OrderModule);
            _Types[typeof(IArchiveRepository)] = typeof(Models.dummy.ArchiveRepository);
        }

        private readonly Dictionary<Type, Type> _Types = new Dictionary<Type, Type>();

        protected T GetObject<T>()
            where T : class
        {
            var targetType = _Types[typeof(T)];
            return targetType.InvokeMember(null, System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.CreateInstance, null, null, null) as T;
        }
    }
}